rust/src/composepost.rs

```diff
@@ -549,10 +552,153 @@ pub fn composepost_nsswitch_altfiles(rootfs_dfd: i32) -> CxxResult<()> {
+            Ok(())
+        },
+    )?;
+
+    Ok(())
+}
+
+/// Recursively explore target directory and translate content to tmpfiles.d entries.
+///
+/// This proceeds depth-first and progressively delete translated subpaths as it goes.
+#[context("Exploring /{} content", prefix)]
+fn convert_path_to_tmpfiles_d_recurse(
+    tmpfiles_bufwr: &mut BufWriter<File>,
+    pwdb: &PasswdDB,
+    rootfs: &openat::Dir,
+    prefix: &str,
+    cancellable: &Option<&gio::Cancellable>,
+) -> Result<()> {
+    use openat::SimpleType;
+
+    for subpath in rootfs.list_dir(prefix)? {
+        if cancellable.map(|c| c.is_cancelled()).unwrap_or_default() {
+            return Ok(());
+        };
+
+        let subpath = subpath?;
+        let path_name = subpath.file_name().to_string_lossy();
+        let full_path = format!("{}/{}", prefix, &path_name);
+        let path_type = subpath.simple_type().unwrap_or(SimpleType::Other);
+
+        // Workaround for nfs-utils in RHEL7:
+        // https://bugzilla.redhat.com/show_bug.cgi?id=1427537
+        let mut retain_entry = false;
+        if path_type == SimpleType::File && full_path.starts_with("var/lib/nfs") {
+            retain_entry = true;
+        }
+
+        if !retain_entry && path_type != SimpleType::Dir && path_type != SimpleType::Symlink {
+            rootfs.remove_file_optional(&full_path)?;
+            println!("Ignoring non-directory/non-symlink '{}'", &full_path);
+            continue;
+        }
+
+        let filetype_char = match path_type {
+            SimpleType::Dir => 'd',
+            SimpleType::Symlink => 'L',
+            SimpleType::File => 'f',
+            x => unreachable!("invalid path type: {:?}", x),
+        };
+        write!(tmpfiles_bufwr, "{} ", filetype_char)?;
+        write!(tmpfiles_bufwr, "/{} ", full_path)?;
+
+        if path_type == SimpleType::Symlink {
+            let link_target = rootfs.read_link(&full_path)?;
+            write!(tmpfiles_bufwr, "- - - - ")?;
+            write!(tmpfiles_bufwr, "{}", link_target.display())?;
+        } else {
+            let meta = rootfs.metadata(&full_path)?;
+            let perm = meta.stat().st_mode & !libc::S_IFMT;
+            write!(tmpfiles_bufwr, "{:04o} ", perm)?;
+            let username = pwdb.lookup_user(meta.stat().st_uid)?;
+            write!(tmpfiles_bufwr, "{} ", username)?;
+            let groupname = pwdb.lookup_group(meta.stat().st_gid)?;
+            write!(tmpfiles_bufwr, "{} ", groupname)?;
+            write!(tmpfiles_bufwr, "- -")?;
+        };
+        write!(tmpfiles_bufwr, "\n")?;
+
+        if path_type == SimpleType::Dir {
+            // New subdirectory discovered, recurse into it.
+            convert_path_to_tmpfiles_d_recurse(
+                tmpfiles_bufwr,
+                pwdb,
+                rootfs,
+                &full_path,
+                cancellable,
+            )
+            .with_context(|| format!("Analyzing {}", &full_path))?;
```

Comment:
I think having `.context()` on recursive calls is a bit of an anti pattern because here it will end up like:
`Analyzing /var: Analyzing /var/lib: Analyzing /var/lib/foo ...`.
In this case I think we need to handle it at each call site, or factor out a helper function that does the inner work and only has context there.

Reply:
Ah that's true, and it also duplicates with function-context at the top.